cmd/peer-rest-client.go

```diff
@@ -423,51 +423,49 @@ func (client *peerRESTClient) SignalService(sig serviceSignal) error {
 }
 
 // Trace - send http trace request to peer nodes
-func (client *peerRESTClient) Trace(doneCh chan struct{}, trcAll bool) (chan []byte, error) {
-	ch := make(chan []byte)
+func (client *peerRESTClient) Trace(traceCh chan interface{}, doneCh chan struct{}, trcAll bool) {
 	go func() {
-		cleanupFn := func(cancel context.CancelFunc, ch chan []byte, respBody io.ReadCloser) {
-			close(ch)
-			if cancel != nil {
-				cancel()
-			}
-			http.DrainBody(respBody)
-		}
 		for {
 			values := make(url.Values)
 			values.Set(peerRESTTraceAll, strconv.FormatBool(trcAll))
+
 			// get cancellation context to properly unsubscribe peers
 			ctx, cancel := context.WithCancel(context.Background())
 			respBody, err := client.callWithContext(ctx, peerRESTMethodTrace, values, nil, -1)
 			if err != nil {
-				//retry
+				// Retry the failed request.
 				time.Sleep(5 * time.Second)
-				select {
-				case <-doneCh:
-					cleanupFn(cancel, ch, respBody)
-					return
-				default:
+			} else {
+				dec := gob.NewDecoder(respBody)
+
+				go func() {
+					<-doneCh
+					// cancel() the request so that bio.Scan() can return.
+					cancel()
+				}()
+
+				for {
+					var info trace.Info
+					if err = dec.Decode(&info); err != nil {
+						break
```

Comment:
Also on error, we need to call `cancel` to not leak the context.

Reply:
it finally gets cleaned up when the doneCh is called:

```
                                go func() {
                                        <-doneCh
                                        // cancel() the request so that bio.Scan() can return.
                                        cancel()
                                }()
```

During the retries on error there will be multiple such goroutines blocked on the `doneCh` but it's OK as they return when ultimately `doneCh` is closed.

Reply:
I have refactored this code into a new function to take advantage of `defer` statement.